### ISSUE TITLE
Support arrays

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/copilot-verifier#7]
We currently have some FIXMEs involving arrays that we should address:

* https://github.com/GaloisInc/copilot-verifier/blob/b7c7a2454200ccec6698b66cf8b02012ad37934b/copilot-verifier/src/Copilot/Verifier.hs#L596-L597
* https://github.com/GaloisInc/copilot-verifier/blob/b7c7a2454200ccec6698b66cf8b02012ad37934b/copilot-verifier/src/Copilot/Verifier.hs#L641-L642

**Comment by RyanGlScott** [comment]
If I try verifying a [simple array example from the `copilot` repo](https://github.com/Copilot-Language/copilot/blob/e3f4e84b496a76f05cb615380f77b672438e127b/copilot/examples/Array.hs):

```hs
--------------------------------------------------------------------------------
-- Copyright © 2019 National Institute of Aerospace / Galois, Inc.
--------------------------------------------------------------------------------

-- | This is a simple example for arrays. As a program, it does not make much
-- sense, however it shows of the features of arrays nicely.

-- | Enable compiler extension for type-level data, necesary for the array
-- length.

{-# LANGUAGE RebindableSyntax #-}
{-# LANGUAGE DataKinds        #-}

module Copilot.Verifier.Examples.Array where

import Language.Copilot
import Copilot.Compile.C99
import Copilot.Verifier (verify)

-- Lets define an array of length 2.
-- Make the buffer of the streams 3 elements long.
arr :: Stream (Array 2 Bool)
arr = [ array [True, False]
      , array [True, True]
      , array [False, False]] ++ arr

spec :: Spec
spec = do
  -- A trigger that fires 'func' when the first element of 'arr' is True.
  -- It passes the current value of arr as an argument.
  -- The prototype of 'func' would be:
  -- void func (int8_t arg[3]);
  trigger "func" (arr .!! 0) [arg arr]

-- Compile the spec
main :: IO ()
main = reify spec >>= verify mkDefaultCSettings [] "array"
-- main = interpret 30 spec
```

Then it actually fails before we ever reach any of the FIXMEs:

```
=====
== Running the array example...
=====

Generated "results/array/array.c"
Compiled array into results/array/crux~array.bc
Translated bitcode into Crucible
Generating proof state data
Computing initial state verification conditions
verify-examples: You have encountered a bug in Crucible's implementation.
*** Please create an issue at https://github.com/GaloisInc/crucible/issues

%< --------------------------------------------------- 
  Revision:  ffa247c549e42ff3325eb8f03fd2eacfc8480a14
  Branch:    master
  Location:  MemModel.unpackMemValue
  Message:   Crucible type mismatch when unpacking LLVM value
             *** Crucible type: IntrinsicRepr LLVM_pointer [BVRepr 1]
             *** LLVM value: <int8>
CallStack (from HasCallStack):
  panic, called at src/Lang/Crucible/Panic.hs:11:9 in crucible-0.5-inplace:Lang.Crucible.Panic
  panic, called at src/Lang/Crucible/LLVM/MemModel.hs:1424:3 in crucible-llvm-0.3-inplace:Lang.Crucible.LLVM.MemModel
  unpackMemValue, called at src/Lang/Crucible/LLVM/MemModel.hs:1407:15 in crucible-llvm-0.3-inplace:Lang.Crucible.LLVM.MemModel
  unpackMemValue, called at src/Lang/Crucible/LLVM/MemModel.hs:577:3 in crucible-llvm-0.3-inplace:Lang.Crucible.LLVM.MemModel
%< ---------------------------------------------------
```

**Comment by RyanGlScott** [comment]
While investigating this, I uncovered something strange about the treatment of `Bool`s in `copilot-verifier`, which may be related to the failure above. `copilotTypeToMemType'` gives special treatment to `Bool`s:

https://github.com/GaloisInc/copilot-verifier/blob/b7c7a2454200ccec6698b66cf8b02012ad37934b/copilot-verifier/src/Copilot/Verifier.hs#L648-L653

And `copilotTypeToMemType` treats `Bool`s differently:

https://github.com/GaloisInc/copilot-verifier/blob/b7c7a2454200ccec6698b66cf8b02012ad37934b/copilot-verifier/src/Copilot/Verifier.hs#L655-L662

Note that the former treats `Bool`s as `i1`, whereas the latter treats `Bool`s as `i8`. Different parts of `copilot-verifier` use `copilotTypeToMemType` versus `copilotTypeToMemType'`, and I'm not sure I understand why each function is used where it is. @robdockins, do you recall the reasoning behind d676293273f15b322fd7d9cf1dd035c4f7fb23b6?

**Comment by robdockins** [comment]
It's some pipe-fitting to do with the fact that the `Bool` type is treated differently in registers and arguments than in memory (where they are extended to a byte).  Clang uses `i1` to handle boolean values in virtual registers, and extends them to `i8` when they get written to memory, and coerces them back somehow (I forget exactly what it does) when read.

**Comment by RyanGlScott** [comment]
Ignoring the issue in https://github.com/GaloisInc/copilot-verifier/issues/7#issuecomment-956291201 for a moment, I've been investigating a different example that doesn't involve `bool`s:

```hs
{-# LANGUAGE RebindableSyntax #-}
{-# LANGUAGE DataKinds        #-}

module Main where

import Copilot.Compile.C99
import Language.Copilot

arr :: Stream (Array 2 Int32)
arr = [ array [0, 1]
      , array [2, 3]
      , array [4, 5]] ++ arr

spec :: Spec
spec = do
  trigger "func" true [arg arr]

main :: IO ()
main = do
  spec' <- reify spec
  compile "array" spec'
```

This isn't quite as straightforward as you might expect. Although `copilot-c99` generates this C code for `func`:

```c
void func(int32_t func_arg0[(2)]);
```

The corresponding LLVM function signature is:

```
declare dso_local void @func(i32*) #2
```

Note that rather than, say, `[2 x i32]`, we have `i32*` as an argument. As a result, we can't just map Copilot `Array` types to `Array` `MemType`s in Crucible, but rather to pointers. After applying some hacks to make that happen, I end up with this error:

```
=====
== Running the array example...
=====

Generated "results/array/array.c"
Compiled array into results/array/crux~array.bc
Translated bitcode into Crucible
Generating proof state data
Computing initial state verification conditions
Proving initial state verification conditions
Proved 3 of 3 total goals
Computing step bisimulation verification conditions
verify-examples: user error (Simulation aborted!
Abort due to assertion failure:
  results/array/array.c:31:6: error: in step
  Mismatch between Copilot value and crucible value
  [XInt32 c@15:bv,XInt32 c@16:bv]
  IntrinsicRepr LLVM_pointer [BVRepr 64]

)
```

This makes sense when you think about it, as the override for `func` is given an `i32*` value in LLVM, but the corresponding Copilot value is an array of two `XInt32`s. It seems like we need to resolve this impedance mismatch by converting the `LLVMPtr` representing the `i32*` value to a vector of two `i32`s, although I'm not sure how to do this off the top of my head.

**Comment by RyanGlScott** [comment]
Supporting structs has its own challenges. For this reason, I've split off the task of supporting structs into a separate issue in #9.

**Comment by robdockins** [comment]
We'll need to push the memory down into the `copilotExprToRegValue` and `computeEqualVals` functions so that we can use the pointer to do indirection to write/read the appropriate values.  The setup phase (`copilotExprToRegValue`) will need to mutate the memory to do this, and maybe it should have a better name.

**Comment by robdockins** [comment]
Humm... well, I was thinking that we would need to write arrays element-by-element, but the memory model is perfectly capable of doing bulk writes of structured values, so I guess you're right that it shouldn't be necessary.

**Comment by RyanGlScott** [comment]
Phew, I got lucky there!

For what it's worth, I also discovered an answer to the question I raised in https://github.com/GaloisInc/copilot-verifier/issues/7#issuecomment-957698495. When an argument is a raw `bool`, LLVM represents it as an `i1`. When a `bool` is contained within a composite type like a pointer, struct, or array, on the other hand, LLVM represents it as an `i8`. It turns out that `copilot-verifier` partially accounts for this distinction by having separate `copilotTypeToMemType'` (which treats `bool` as an `i8`) and `copilotTypeToMemType` (which treats `bool` as an `i1`) functions. It doesn't go far enough, however, as the code which generates `Array` types should be using `copilotTypeToMemType'`, not `copilotTypeToMemType`.

I'll polish up what I have and, with luck, I'll have a PR ready tomorrow.
